pyroscope-io/grafana
# [WIP] pyroscope flamechart Explore

Here is basic logic of Pyroscope Flamechart support on Explore.

Demo video: https://monosnap.com/file/ggZOwKjvLYKjqURJ5P9V4iQHmNDOtj

How to reproduce it locally:

* clone the repo or brach pyroscope-flamechart-explore
* checkout to this branch
* start Pyroscope instance (I did it as docker container docker run -it -p 4040:4040 pyroscope/pyroscope:latest server)
* start Grafana: `make build`, then `bin/darwin-amd64/grafana-server`
* add ppsPlugin as datasource and chose it as datasource at the top of Explore page
* execute some Pyroscope query

Some thoughts on this POC. There is a chance that I made anything in wrong way. Please let me know what do you think about that. Any thougts and planning my further work would be appreciated.

* What is ppsPlugin? I moved pyroscope-datasouce to Grafana as built-in plugin just for test purposes. It'll be deleted in final version of this PR (all code under public/app/plugins/datasource/pyroscope). But you need to change preferredVisualisationType to profile instead of table in packages/pyroscope-datasource-plugin/src/datasource.ts

* I added pyroscope-panel + pyroscope-flamegraph as built-in plugin for Grafana. As I understand there is no way to adopt pyroscope-panel plugin to work as visualization module on Explore except setting it as built-in plugin. Code, which I moved there looks not really graceful, I had to cut a lot of code to reduce code size and focus on rendering flamegraphs. Almost all code has been moved from Pyroscope plugins as is. Grafana doesn't work with css modules, so I had to set some styles in inline way (<div style={{display: 'flex'}} />). and ts-lint has different rules so I decorated some files and lines by @ts-nocheck and @ts-ignore.

* Injecting pyroscope-flamegraph as instant plugin assumes extending Grafana package.json deps. I had to add these packages to project:
@react-hook/resize-observer
@szhsin/react-menu
color
@types/color
d3-scale
@types/d3-scale
clsx
true-myth
ts-essentials

* Grafana has built-in plugin TestData DB, which emulates different data types and how they handled on Explore page. I think we should also add Pyroscope data format support for this plugin. Guys who added support for nodeGraph did it in this way.

---

Copied from https://github.com/pavelpashkovsky/grafana/pull/2
